hubble/relay: improve peer connections handling

rolinh opened the pull request:
This PR improves peer connections handling and introduce unit tests. In order to implement unit tests, a major refactoring was necessary. This refactoring also has for side-effect to improve separation of concerns as all the peer connection handling logic is now into its own package (`pkg/hubble/relay/pool`).

For more details, please go through the list of commits.

Rel: #11425

rolinh committed:
hubble/relay: decouple peer syncing logic from the server struct

This is a first step that will allow implementing unit tests for the
relay package. A new interface, `peerSyncer`, is defined and implemented
by a `syncer` struct. Having the interface will allow creating a mock
for testing purposes.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: move reconnection logic to peer syncer

This provides a better abstraction as the client that calls `List()` is
expected to receive a list of peers that are "ready" to connect to and
does not have to manage re-connect.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: add a pool manager to handle peers connections

This commit is an in-depth reimplementation of the peers
synchronization/connections handling logic. All logic regarding
management of peers and connections to them has been moved to a separate
package with a clearly defined interface for the consumer
(`pool.PeerManager`).

The goal here is threefold:

  1. Separation of concerns (by hiding all internal peer handling logic
     into a dedicated package)
  2. Increased testability
  3. Improved connections handling

The consumer of the `pool` package does not have to care whether a peer
is connected or not anymore. If a peer is not connected, it's `Conn`
attribute will be `<nil>`. If `Conn` is not `<nil>` but the peer cannot
be reached, the consumer can signal it by using the manager's
`ReportOffline(name string)` method.

The new `pool.Manager` (which implements the `pool.PeerManager`
interface) also implements a better mechanism to ensure peers connection
are healthy by using a dedicated routine that monitors for
offline/reported offline peers and attempts to (re-)connect to them.
This improves the situation as with the older implementation,
re-connecting to a peer required a new call to the Hubble Relay service
(and the peers that were "woken up" were not available for the request).

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/peer: define a Peer service Client interface and Builder

Such interface are useful to create good abstractions with code that
require a Peer service client. They can also easily be mocked which is
essential to write good unit tests.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/peer: implement ClientBuilder for a local connection

LocalClientBuilder creates a new Client which is suitable when
the gRPC connection to the Peer service is local (typically a Unix
Domain Socket).
This scenario is the most common one and is directly useful to Hubble
Relay.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: use peer.ClientBuilder to build Peer service clients

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: define and use a gRPC client connection builder

This commits defines an interface for a gRPC client connection builder.
When instantiating a new pool.Manager, the caller may specify a grpc
client connection builder implementation to be used to create connection
to Hubble peers. This allows for fine grained control over the client
connection parameters by the caller. In addition, as this is an
interface, it can easily be mocked which is useful when writing unit
tests.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: implement backoff for peer connection attempts

With the rework of peer connections handling, a goroutine periodically
checks whether peer connections are healthy or not. This check is
necessary as there are scenarios where a connection can be lost without
any peer service notification or client request (for instance, a node
reboot).

This commit improves reconnection attempts by implementing a backoff
mechanism so attempts are not too frequent. There is a default backoff
mechanism that can be configured by caller when instantiating a new
pool.Manager.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: make the connection check interval configurable

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: report offline peers to the pool manager

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: ignore backoff on direct connection requests

On peer change notifications or call to
`pool.Manager.ReportOffline(name string)`, the backoff interval should
be ignored to re-attempt a connection. This change typically addresses
the following scenario:

  1. Node Foo is offline
  2. The connectivity checker attempts to reconnect several time; each
    time the backoff is increased.
  3. Node Foo becomes available; a peer change notification is received
  4. Oops, no connection attempt is made to Foo because of a large
     backoff duration

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: do not call disconnect before connection attempt

This is not necessary as the connect function closes pre-existing
connections.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble: move FakePeerNotifyServer to testutils

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble: move peer interface definitions to hubble/peer/types

This is necessary to avoid import cycles when implementing fakes for the
peer interfaces in the `testutils` package.

Note that the `types` sub-package is commonly found throughout Cilium's
codebase so this change follows the convention.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/testutils: implement fakers for peer/types/{Client,ClientBuilder}

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/testutils: add FakeGRPCClientStream

This will allow mocking gRPC clients which is useful for unit testing.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/testutils: implement FakePeerNotifyClient

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: implement unit tests for the peer pool manager

These unit tests are not exhaustive although they still provide a decent
coverage. However, it provides a good structure to implement more tests
in the future such as complex peer connectivity issue scenarios or
simply regression tests for bugs that are fixed.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

glibsm commented on pkg/hubble/relay/peer.go:
```diff
@@ -40,112 +41,150 @@ func newPeerClient(target string, dialTimeout time.Duration) (peerpb.PeerClient,
+	Start()
+	Stop()
+	List() []hubblePeer
+	ConnectPeer(name, address string)
```
interface is private, yet fields are public. what's the logic there?

rolinh replied:
This was part of the first refactoring commit, I wasn't sure at the time what the right interface should be. The corresponding interface is now and everything is public because it's in its own package (`pkg/hubble/peer/pool`):
```go
// PeerManager defines the functions a peer manager must implement when
// handling peers and respective connections.
type PeerManager interface {
	// Start instructs the manager to start peer change notification handling
	// and connection management.
	Start()
	// Stop stops any peer manager activity.
	Stop()
	// List returns a list of peers with active connections. If a peer cannot
	// be connected to; its Conn attribute must be nil.
	List() []Peer
	// ReportOffline allows the caller to report a peer as being offline. The
	// peer is identified by its name.
	ReportOffline(name string)
}
```

glibsm commented on pkg/hubble/relay/pool/option.go:
```diff
@@ -0,0 +1,77 @@
+	DialTimeout  time.Duration
+	RetryTimeout time.Duration
+	Debug        bool
+}
```
i don't think any of these have to be public

rolinh replied:
They don't _have to_ be for documentation purpose it makes sense so that the user can see what the defaults are. Current version is:
```go
// DefaultOptions is the reference point for default values.
var DefaultOptions = Options{
        PeerClientBuilder: peerTypes.LocalClientBuilder{
                LocalAddress: "unix://" + defaults.HubbleSockPath,
                DialTimeout:  5 * time.Second,
        },
        ClientConnBuilder: GRPCClientConnBuilder{
                DialTimeout: 5 * time.Second,
                Options: []grpc.DialOption{
                        grpc.WithInsecure(),
                        grpc.WithBlock(),
                },
        },
        Backoff: &backoff.Exponential{
                Min:    10 * time.Second,
                Max:    12 * time.Hour,
                Factor: 2.0,
        },
        ConnCheckInterval: 2 * time.Minute,
        RetryTimeout:      30 * time.Second,
}

// Option customizes the configuration of the Manager.
type Option func(o *Options) error

// Options stores all the configuration values for peer manager.
type Options struct {
        PeerClientBuilder peerTypes.ClientBuilder
        ClientConnBuilder ClientConnBuilder
        Backoff           BackoffDuration
        ConnCheckInterval time.Duration
        RetryTimeout      time.Duration
        Debug             bool
}
```

glibsm commented on pkg/hubble/peer/client.go:
```diff
@@ -0,0 +1,35 @@
+// ClientBuilder creates a new Client.
+type ClientBuilder interface {
+	// Target returns the address that Client connects to.
+	Target() string
```
builder would be `Target(t string) ClientBuilder`, no?

glibsm replied:
maybe I was thinking about another pattern where you chain all of them together `Foo.Target(name).Port(123).Build()`

rolinh replied:
This interface is named `ClientBuilder` because it "builds" a `Client` with the `Client()` method. I'm ok to change the name if its confusing but I couldn't come up with a better one at the time.

gandro replied:
I also stumbled on this. 

I think the un-intuitiveness is less a naming problem and comes from the fact typically a factory returns objects with different configurations. In this case, the interface assumes that the configuration is hard-coded and all returned clients share the exact same configuration (i.e. the same target).

Would it be possible to keep the configuration and the factory separate? I.e. modify the method below to `Client(target string) (Client, error)` and keep the `WithTarget` server option? 

Fake `ClientBuilder` implementations would simply ignore the target, while the real one will use it to connect to the local unix domain socket (or throw an error if the target is not supported by that `ClientBuilder` implementation).

Alternatively: Make `Target()` a function of the `Client` interface. Then the `ClientBuilder` does not specify if the target is hard-coded (as the current version) or configurable (my proposal), since you are only using it to log properties of the client (and not the builder).

rolinh replied:
OK, I see your point now. So your suggesting this interface to basically be similar to the `ClientConnBuilder`:
```go
// ClientConnBuilder wraps the ClientConn method.
type ClientConnBuilder interface {
        // ClientConn creates a new ClientConn using target.
        ClientConn(target string) (ClientConn, error)
}
```

The reason I made the `ClientBuilder` interface different is because, as opposed to the `ClientConnBuilder` where each peer address is known by the peers manager, the peer service address is unknown unless another option to provide the peer service target is added. So the `Target()` method was added to the interface mainly to be able to log the peer service address in case of connection problems. At the time, I also thought that having two options to set define the way to connect to the peer service would be error prone. But seeing confusion, I'm also fine updating this interface so the two look similar.

rolinh replied:
`Target()` has been removed from the interface and `Client` now takes `target string` as parameter.

gandro commented on pkg/hubble/relay/pool/option.go:
```diff
@@ -35,6 +37,11 @@ var DefaultOptions = Options{
 	},
+	Backoff: &backoff.Exponential{
+		Min:    10 * time.Second,
+		Max:    12 * time.Hour,
```
Max of 12 hours seems excessive to me. I would rather give up at some point than waiting for 12 hours to see if it comes back up. 

It seems that we're not using a maximum often in Cilium (most uses don't seem specify it at all (i.e. unbounded), and all others seem below 2 minutes). I personally would have set it it somewhere in the range of 20-90 minutes at most, but that's also just a random number.

rolinh replied:
Note that this backoff is exponential. So with these settings, one gets this result:
```
10s
20s
40s
1m20s
2m40s
5m20s
10m40s
21m20s
42m40s
1h25m20s
2h50m40s
5h41m20s
11h22m40s
12h0m0s
12h0m0s
12h0m0s
12h0m0s
12h0m0s
12h0m0s
12h0m0s
```
You can `go run` this code to test with different values:
```go
package main

import (
        "fmt"
        "time"

        "github.com/cilium/cilium/pkg/backoff"
)

func main() {
        b := backoff.Exponential{
                Min:    10 * time.Second,
                Max:    12 * time.Hour,
                Factor: 2.0,
        }

        for i := 0; i < 20; i++ {
                fmt.Println(b.Duration(i))
        }
}
```

I'm not sure what the behavior should be. Maybe we should just give up after a certain number of attempts and mark the node as permanently unavailable?

gandro replied:
> I'm not sure what the behavior should be. Maybe we should just give up after a certain number of attempts and mark the node as permanently unavailable?

Yeah. What I'm proposing is at least to make the function stop way earlier. I don't see the value in waiting hours between reconnects.

```
10s
20s
40s
1m20s
2m40s
5m20s
10m40s
21m20s
42m40s
60m0s
60m0s
60m0s
```

Additionally giving up after a few hours (e.g. 10-20 retries or so) seems reasonable to me.

rolinh replied:
I set max to 90mn. In any case, this is just a default and can be configured by the caller.

gandro commented on pkg/hubble/relay/pool/option.go:
```diff
@@ -42,7 +42,8 @@ var DefaultOptions = Options{
 		Factor: 2.0,
 	},
-	RetryTimeout: 30 * time.Second,
+	ConnCheckInterval: 2 * time.Minute,
```
How does this interact with the backoff? I think if we only trigger reconnects every `ConnCheckInterval`, then the minimal backoff will essentially always be at least `ConnCheckInterval`, in this case 2 minutes. The first retry which is supposed to happen after 10 seconds according to exponential backoff will actually happen after 2 minutes.

In addition, we will start all re-connection attempts always at the same time for all offline peers whose backoff has expired in the last 2 minutes. There is no jitter.

I'm not sure what the best approach is and we can do it in a follow-up PR. But is seems to me that only reconnecting all peers `ConnCheckInterval` will lead to very bursty reconnection behavior. 

Ideally, we'd have an interval for checking the connection status, and a a separate logic scheduling the reconnection (i.e. a timer which expires at the smallest `nextConnAttempt`)

rolinh replied:
> I think if we only trigger reconnects every `ConnCheckInterval`

Reconnects are also triggered with a client requests that uses `ReportOffline`.

> The first retry which is supposed to happen after 10 seconds according to exponential backoff will actually happen after 2 minutes.

Unless there's a client request, yes, this is true. I'm not sure it's worth trying to reconnect more aggressively than this.

> In addition, we will start all re-connection attempts always at the same time for all offline peers whose backoff has expired in the last 2 minutes. There is no jitter.

Very good point. I thought about this as well but as this PR was already getting pretty big, I thought this could be addressed in a follow-up PR. Note that even before this PR there's also anbother burst when Hubble Relay is started and is notified of all the peers in the cluster.

I fully agree with your point here and I'm happy to discuss a solution to implement in a follow-up PR.

gandro replied:
> Unless there's a client request, yes, this is true. I'm not sure it's worth trying to reconnect more aggressively than this.

If there is a client reconnect, then the backoff is ignored anwyway. Currently, actual back-off function will look like this. 

 ```
2m0s (ConnCheckInterval)
2m0s (ConnCheckInterval)
2m0s (ConnCheckInterval)
2m0s (ConnCheckInterval)
2m40s
5m20s
10m40s
21m20s
42m40s
1h25m20s
2h50m40s
5h41m20s
11h22m40s
12h0m0s
```

I agree that we do not have to fix this in this PR, but it seems like this is not intentional.

gandro commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -178,7 +178,8 @@ func (m *Manager) manageConnections() {
 			m.mu.Unlock()
-			m.connect(p)
+			// a connection request has been made, make sure to attempt a connection
+			m.connect(p, true)
```
Note to reviewers, since it's not clear from context: `m.connect()` will not reconnect if there is already a re-connection attempt in progress. So it's fine if the same peer is reported offline multiple times, we will still only do one actually re-connection attempt in `m.connect`.

gandro commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -0,0 +1,632 @@
+									default:
+										return cns[i], nil
+									}
+								},
```
nit (no change required): 

I think readability of the tests would benefit greatly if we would could extract the common logic of the different `FakePeerClient` implementations here. They are all similar, but also slightly different. Plus due to the fact that we have nested `func()`s and non-local effects (access to the `done` variable which is defined and accessed hundreds of lines of code away) makes code is also not easily readable.

Now I'm actually not sure if there is common logic that can really be extracted. So again: No change required. Better to have hard-to-read unit tests than no unit tests at all. This code provides a ton of value already!
But if there was a helper to implement custom `FakePeerClient` to be extracted without much effort here, that would make future maintenance of these unit tests much easier.

rolinh replied:
I agree these tests are not the easier to read and it's mostly due to having three layers of mocked interfaces. I'm not really sure how could this be addressed besides adding helpers for the most common mock configs.

gandro commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -0,0 +1,632 @@
+				},
+			},
+		}, {
+			name: "PeerClientBuilder errors out",
```
Maybe I'm missing something, but it seems we are not checking if these errors are actually thrown somewhere? We just assert that the peer list is empty?

Would it be possible to grep for the error messages in the logger?

rolinh replied:
Indeed. My goal with this test was mostly to make sure this does not lead to a crash. I actually think that it would make sense to allow passing a logger as an option so one could also use a customer logger ([`logrus.FieldLogger`](https://godoc.org/github.com/sirupsen/logrus#FieldLogger) is an interface) for tests that would log to a buffer such that one can assert on the log output.

rolinh replied:
I've updated the tests for test the logger's output.

gandro reviewed (commented):
:tada: This is an awesome change! The resulting code looks much cleaner and the fact that we can now easily mock connectivity issues and faulty client responses is simply amazing.

The implementation itself looks solid. I do have a few (mostly minor) higher-level questions and comments that popped up during the review that I wanted to discuss before approving.

rolinh committed:
hubble/[peer|relay]: remove Target() from the ClientBuilder interface

This commit removes `Target()` from the peer `ClientBuilder` interface
and, instead, adds the target as a parameter to `Client()`.

As a consequence, Hubble Relay code is adapted to the change. The peer
pool manager now has a new option to provide the address of the peer
gRPC service.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: add pool option to set the logger and remove debug option

The debug option was only used to set the log level of the logger. As
the caller now has control over the logger to be used, this option is no
longer necessary.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

aanm commented on pkg/hubble/relay/observer.go:
```diff
@@ -32,16 +33,27 @@ import (
 	"google.golang.org/grpc/status"
 )
 
+// TODO: remove this shim once "google.golang.org/grpc" is bumped to v1.28+
```
Why don't we bump it now?

rolinh replied:
Because bumping this dep breaks `github.com/etcd-io/etcd` (see https://github.com/etcd-io/etcd/pull/11564) which we'd need to bump as well and as these changes need to be backported; it's a bit risky.

aanm replied:
I wasn't aware this was supposed to be backported. :+1:

aanm commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -0,0 +1,361 @@
+	conn            ClientConn
+	connAttempts    int
+	nextConnAttempt time.Time
+	mu              lock.Mutex
```
what's this mutex protecting?

rolinh replied:
All `peer` struct members.

aanm replied:
>When declaring a struct where the mutex must protect access to one or more fields, place the mutex above the fields that it will protect as a best practice.

rolinh replied:
Where does this quote come from? In any case, I've reordered the struct members.

aanm commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -0,0 +1,361 @@
+type Manager struct {
+	peers   map[string]*peer
+	offline chan string
+	mu      lock.Mutex
```
Is this mutex protecting `opts` and `stop`?

rolinh replied:
Options are set once when the `Manager` is created. This mutex protects the `peers` map.

aanm replied:
>When declaring a struct where the mutex must protect access to one or more fields, place the mutex above the fields that it will protect as a best practice

aanm commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -0,0 +1,361 @@
+	return &Manager{
+		peers:   make(map[string]*peer),
+		offline: make(chan string, 100),
+		stop:    make(chan struct{}),
```
This `stop` couldn't this stop be context? [1]

rolinh replied:
It _could_ be replaced by a context but I try to follow the best practice documented in the [`context` package documentation](https://golang.org/pkg/context/#pkg-overview):
>  Programs that use Contexts should follow these rules to keep interfaces consistent across packages and enable static analysis tools to check context propagation:
>
> **Do not store Contexts inside a struct type**; ...

aanm commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -0,0 +1,361 @@
+}
+
+func (m *Manager) watchNotifications() {
+	ctx, cancel := context.WithCancel(context.Background())
```
[1] then it could be used as a parent of this context.

aanm commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -0,0 +1,361 @@
+					}
+				}
+				if p.nextConnAttempt.IsZero() || p.nextConnAttempt.Before(now) {
+					retry = append(retry, p)
```
Given that `m.connect` has a go routine internally why don't we run `go m.connect(p, false)` in here instead of storing all peers that we need to retry in the `retry` slice?

rolinh replied:
Good point. Initially the code was a bit more complicated so I wanted to move the logic out of the `m.mu.Lock()` block. But I guess I can simplify this now.

aanm commented on pkg/hubble/relay/observer.go:
```diff
@@ -250,20 +262,20 @@ func (s *Server) GetFlows(req *observerpb.GetFlowsRequest, stream observerpb.Obs
 
 	for _, p := range peers {
 		p := p
```
What's this line doing?

rolinh replied:
See [Using reference to loop iterator variable](https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable).

aanm replied:
I didn't see this was used inside `g.Go(func() error {`

aanm commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -0,0 +1,361 @@
+type Manager struct {
+	peers   map[string]*peer
+	offline chan string
+	mu      lock.Mutex
```
I might be wrong but this mutex could be replaced with a `lock.RWMutex`

rolinh replied:
Same as above, I fail to see what benefit this change would bring.

aanm replied:
It would allow to have multiple calls of the `List() []Peer` function, for example.

rolinh replied:
I reviewed the code carefully and it indeed seems that we expect many more reads that writes. I've changed the mutex type to be a `RWMutex`.

aanm commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -0,0 +1,361 @@
+
+// ReportOffline implements PeerManager.ReportOffline.
+func (m *Manager) ReportOffline(name string) {
+	go func() {
```
The creation of the go routine should be done by the caller to be more explicit that the `ReportOffline` is going to be done async.

rolinh replied:
I think that the caller should not care about this implementation detail.

aanm commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -0,0 +1,361 @@
+		if p.conn != nil {
+			switch p.conn.GetState() {
+			case connectivity.Connecting, connectivity.Idle, connectivity.Ready:
+				// it looks like it's actually online or being brought online
```
How do we know that? What guarantees do we have that this go routine was not concurrently scheduled with a 
 `m.connect(p, ...)`?

rolinh replied:
It may have been but it doesn't really matter; this is more a sanity check to ensure the caller does not report offline a peer that is indeed online (or about to become) to avoid unnecessary work. Also, there's a lock on `p`.

aanm commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -0,0 +1,361 @@
+	conn            ClientConn
+	connAttempts    int
+	nextConnAttempt time.Time
+	mu              lock.Mutex
```
I might be wrong but this mutex could be replaced with a `lock.RWMutex`

rolinh replied:
Why would we do this? I think it would be detrimental to performance in this case.

rolinh replied:
I don't think we would benefit from using a `RWMutex` in this specific case.

aanm commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -0,0 +1,361 @@
+}
+
+func (m *Manager) connect(p *peer, ignoreBackoff bool) {
+	go func() {
```
The creation of the go routine should be done by the caller to be more explicit that the `connect` is going to be done async.

rolinh replied:
Sure, I'll change this.

aanm commented on pkg/hubble/relay/pool/option.go:
```diff
@@ -0,0 +1,125 @@
+	ClientConnBuilder: GRPCClientConnBuilder{
+		DialTimeout: 5 * time.Second,
+		Options: []grpc.DialOption{
+			grpc.WithInsecure(),
```
I assume this also requires a `grpc.WithInsecure(), // FIXME: remove once mTLS is implemented`?

rolinh replied:
I think this case is fine assuming the connection is local on a Unix Domain Socket (as is the default).

rolinh committed:
hubble/relay: wait on goroutine to finish when pool.Stop() is called

This change makes tests more reliable and is cleaner at the same time as
once the function returns, the pool manager is really shut down.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: assert logger messages in pool manager tests

This commit tests the pool manager logger output for certain tests. This
change revealed some flakiness in the tests that have been addressed.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

gandro commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -53,10 +60,8 @@ func TestManager(t *testing.T) {
-										close(done)
-									})
-									return nil, nil
+									defer func() { once.Do(func() { close(done) }) }()
```
Why a defer right before a return? Maybe I'm missing something, but a `once.Do(func() { close(done) })` would do the trick here as well?

gandro commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -86,12 +91,8 @@ func TestManager(t *testing.T) {
-											close(done)
-										}()
-										fallthrough
+										defer func() { once.Do(func() { close(done) }) }()
```
Defer before return?

gandro commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -158,24 +161,24 @@ func TestManager(t *testing.T) {
-							once.Do(func() {
-								close(done)
-							})
+							defer func() { once.Do(func() { close(done) }) }()
```
Defer before return?

gandro commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -272,24 +274,24 @@ func TestManager(t *testing.T) {
-							once.Do(func() {
-								close(done)
-							})
+							defer func() { once.Do(func() { close(done) }) }()
```
Defer before return?

gandro commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -457,19 +471,25 @@ func TestManager(t *testing.T) {
 			},
 			ccBuilder: FakeClientConnBuilder{
 				OnClientConn: func(target string) (ClientConn, error) {
+					defer func() { close(done) }()
```
Defer before return?

gandro commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -479,9 +499,7 @@ func TestManager(t *testing.T) {
-							once.Do(func() {
-								close(done)
-							})
+							defer func() { once.Do(func() { close(done) }) }()
```
Defer before return?

gandro commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -499,9 +522,7 @@ func TestManager(t *testing.T) {
-									once.Do(func() {
-										close(done)
-									})
+									defer func() { once.Do(func() { close(done) }) }()
```
Defer before return?

gandro reviewed (commented):
The new commits look  fine to me in terms of approach! Thanks for addressing my nits and adding some additional checks in the unit tests.

I don't understand the point of the `defer`s though. They seem to try to achieve something, but I don't understand what.

rolinh committed:
hubble/relay: use a RWLock to protect the peers map in the pool Manager

The peers map is expected to be more frequently read than written to,
especially with the `List()` method being exposed and used by Hubble
Relay for every request it handles. Therefore, a RWLock is more
approriate than a regular lock.

Suggested-by: André Martins <andre@cilium.io>
Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

kaworu commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -0,0 +1,378 @@
+	}
+	return &Manager{
+		peers:   make(map[string]*peer),
+		offline: make(chan string, 100),
```
Care to elaborate the rationale behind `100` please?

rolinh replied:
Big enough to absorb a burst, small enough to not occupy to much memory. Also, if the consuming routine can't keep up with the offline notifications; then something is probably wrong somewhere.

kaworu commented on pkg/hubble/relay/server.go:
```diff
@@ -60,11 +53,32 @@ func NewServer(options ...relayoption.Option) (*Server, error) {
 	logger := logging.DefaultLogger.WithField(logfields.LogSubsys, "hubble-relay")
 	logging.ConfigureLogLevel(opts.Debug)
+
+	pool, err := pool.NewManager(
```
nit: I find a bit confusing to declare a local var `pool` when `github.com/cilium/cilium/pkg/hubble/relay/pool` is imported.

rolinh replied:
I renamed it to `pm`.

kaworu commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -0,0 +1,378 @@
+	ReportOffline(name string)
+}
+
+// Peer is like peer.Peer but includes a Conn attribute to reach the peer's
```
```suggestion
// Peer is like hubblePeer.Peer but includes a Conn attribute to reach the peer's
```

rolinh replied:
Nice catch.

kaworu commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -0,0 +1,378 @@
+}
+
+// Manager implements the PeerManager interface.
+type Manager struct {
```
Any particular reason `Manager` is public? My understanding is that exposing the `PeerManager` interface should be enough.

rolinh replied:
`Manager` is an implementation of the `PeerManager` interface and it needs to be public because this is what the constructor returns. See also [CodeReviewComments#interfaces](https://github.com/golang/go/wiki/CodeReviewComments#interfaces):
> The implementing package should return concrete (usually pointer or struct) types: that way, new methods can be added to implementations without requiring extensive refactoring.

kaworu reviewed (changes requested):
Great patch! I only have some minor comments and questions.

rolinh committed:
hubble/relay: rename struct member from pool to pm to avoid confusion

The name `pool` clashes with the package `pool`. It is not a problem per
se but creates a bit of confusion so rename it to `pm`.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

kaworu commented on pkg/hubble/relay/pool/client.go:
```diff
@@ -0,0 +1,67 @@
+// ClientConn is an interface that defines the functions clients need to
+// perform unary and streaming RPCs. It is implemented by *grpc.ClientConn.
+type ClientConn interface {
+	// GetState returns the connectivity.State of ClientConn.
```
Even in grpc v1.30.0 (the latest at the time of writing) `GetState()` is [marked as experimental](https://pkg.go.dev/google.golang.org/grpc?tab=doc#ClientConn.GetState).

Since this PR make use of it in several places, would it make sense to keep track of this somewhere?

rolinh replied:
I think it's important to pay attention to this when upgrading the go-grpc dep but we will anyway have to face breaking changes when upgrading.
